File: app/src/main/java/com/werockstar/reactiveandroid/presenter/GithubPresenter.java
package com.werockstar.reactiveandroid.presenter;


import com.werockstar.reactiveandroid.api.RxApi;
import com.werockstar.reactiveandroid.model.GithubUserResponse;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import retrofit2.HttpException;

public class GithubPresenter {

    private RxApi api;
    private GithubPresenter.View view;

    private final CompositeDisposable disposable = new CompositeDisposable();
    private final String TAG = GithubPresenter.class.getSimpleName();

    @Inject
    public GithubPresenter(RxApi api) {
        this.api = api;
    }

    public interface View {
        void onUsersResult(List<GithubUserResponse> users);

        void onUsersError();
    }

    public void injectView(GithubPresenter.View view) {
        this.view = view;
    }

    public void getUsers() {
        Observable<GithubUserResponse> werockstarObs = api.getUser("werockstar").subscribeOn(Schedulers.io());
        Observable<GithubUserResponse> googleObs = api.getUser("google").subscribeOn(Schedulers.io());
        Observable<GithubUserResponse> facebookObs = api.getUser("facebook").subscribeOn(Schedulers.io());
        Observable<GithubUserResponse> airbnbObs = api.getUser("airbnb").subscribeOn(Schedulers.io());
        Observable<GithubUserResponse> microsoftObs = api.getUser("microsoft").subscribeOn(Schedulers.io());

        disposable.add(Observable.zip(werockstarObs, googleObs, facebookObs, airbnbObs,
                microsoftObs, (w, f, g, a, m) -> {
                    final List<GithubUserResponse> users = new ArrayList<>();
                    users.add(w);
                    users.add(f);
                    users.add(g);
                    users.add(a);
                    users.add(m);
                    return users;
                })
                .onErrorReturn(__ -> {
                    List<GithubUserResponse> users = new ArrayList<>();
                    users.add(new GithubUserResponse("ไม่มีนะ", "นี่ก็ไม่มี", "และนี่ก็ไม่มี"));
                    return users;
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .retry((attempt, throwable) -> {
                    HttpException exception = (HttpException) throwable;
                    return exception.code() != 403 && attempt <= 3;
                })
                .subscribe(
                        users -> {
                            view.onUsersResult(users);
                        }, throwable -> {
                            view.onUsersError();
                        }
                )
        );


    }

    public void dispose() {
        disposable.clear();
    }
}

File: app/src/main/java/com/werockstar/reactiveandroid/presenter/GithubPresenter.kt
package com.werockstar.reactiveandroid.presenter


import com.werockstar.reactiveandroid.api.RxApi
import com.werockstar.reactiveandroid.model.GithubUserResponse
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.Observables
import io.reactivex.rxkotlin.addTo
import io.reactivex.schedulers.Schedulers
import retrofit2.HttpException
import java.util.*
import javax.inject.Inject

class GithubPresenter @Inject constructor(private val api: RxApi) {

    private lateinit var view: GithubPresenter.View
    private val disposable by lazy { CompositeDisposable() }

    interface View {
        fun onUsersResult(users: List<GithubUserResponse>)

        fun onUsersError()
    }

    fun injectView(view: GithubPresenter.View) {
        this.view = view
    }

    fun getUsers() {
        val werockstarObs = api.getUser("werockstar").subscribeOn(Schedulers.io())
        val googleObs = api.getUser("google").subscribeOn(Schedulers.io())
        val facebookObs = api.getUser("facebook").subscribeOn(Schedulers.io())
        val airbnbObs = api.getUser("airbnb").subscribeOn(Schedulers.io())
        val microsoftObs = api.getUser("microsoft").subscribeOn(Schedulers.io())

        Observables.zip(werockstarObs, googleObs, facebookObs, airbnbObs, microsoftObs) { w, f, g, a, m ->
            listOf<GithubUserResponse>(w, f, g, a, m)
        }.onErrorReturn {
            listOf(GithubUserResponse("ไม่มีนะ", "นี่ก็ไม่มี", "และนี่ก็ไม่มี"))
        }.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .retry { attempt, throwable ->
                    val exception = throwable as HttpException
                    exception.code() != 403 && attempt <= 3
                }
                .subscribe({
                    view.onUsersResult(it)
                }, {
                    view.onUsersError()
                }).addTo(disposable)
    }

    fun dispose() {
        disposable.clear()
    }
}
